File: mongodb/concurrency_test.go
package mongodb

import (
	"context"
	"fmt"
	"strings"
	"sync"
	"sync/atomic"
	"testing"
	"time"

	"github.com/DEEJ4Y/scheduler"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

// TestDistributedLocking validates that jobs execute exactly once under high concurrency.
// This test simulates 100 concurrent scheduler instances processing 10,000 jobs.
func TestDistributedLocking(t *testing.T) {
	if testing.Short() {
		t.Skip("Skipping concurrency test in short mode")
	}

	const (
		numSchedulers = 100   // Number of concurrent scheduler instances
		numJobs       = 10000 // Number of jobs to process
		testTimeout   = 5 * time.Minute
	)

	ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
	defer cancel()

	// Connect to MongoDB
	mongoURI := "mongodb://localhost:27017"
	client, err := mongo.Connect(ctx, options.Client().ApplyURI(mongoURI))
	if err != nil {
		t.Skipf("Skipping test: MongoDB not available: %v", err)
	}
	defer client.Disconnect(ctx)

	// Verify connection
	if err := client.Ping(ctx, nil); err != nil {
		t.Skipf("Skipping test: Cannot ping MongoDB: %v", err)
	}

	// Use a unique database for this test
	dbName := fmt.Sprintf("scheduler_concurrency_test_%d", time.Now().Unix())
	db := client.Database(dbName)
	defer db.Drop(context.Background())

	collection := db.Collection("jobs")

	// Create index for performance
	indexModel := mongo.IndexModel{
		Keys:    bson.D{{Key: "sleepUntil", Value: 1}},
		Options: options.Index().SetSparse(true),
	}
	_, err = collection.Indexes().CreateOne(ctx, indexModel)
	if err != nil {
		t.Fatalf("Failed to create index: %v", err)
	}

	t.Logf("Test configuration: %d schedulers, %d jobs, timeout: %v", numSchedulers, numJobs, testTimeout)

	// Job execution tracking
	executions := &ExecutionTracker{
		counts:     make(map[string]int),
		timestamps: make(map[string][]time.Time),
	}

	// Insert jobs
	t.Log("Inserting jobs...")
	startInsert := time.Now()
	if err := insertJobs(ctx, collection, numJobs); err != nil {
		t.Fatalf("Failed to insert jobs: %v", err)
	}
	t.Logf("Inserted %d jobs in %v", numJobs, time.Since(startInsert))

	// Start schedulers
	t.Logf("Starting %d concurrent schedulers...", numSchedulers)
	startTime := time.Now()

	var (
		schedulers []*scheduler.Scheduler
		wg         sync.WaitGroup
		errorCount atomic.Int64
		startMutex sync.Mutex
		stopping   atomic.Bool // Flag to track when shutdown begins
	)

	// Create and start all schedulers
	for i := 0; i < numSchedulers; i++ {
		schedulerID := i

		store, err := NewStore(Config{
			Collection: collection,
		})
		if err != nil {
			t.Fatalf("Failed to create store: %v", err)
		}

		sched, err := scheduler.New(scheduler.Config{
			Store: store,
			OnDocument: func(ctx context.Context, job *scheduler.Job) error {
				// Track execution
				jobID := getJobID(job)
				executions.Record(jobID)
				return nil
			},
			OnError: func(ctx context.Context, err error) {
				// Ignore context canceled errors during shutdown - they're expected
				if stopping.Load() && strings.Contains(err.Error(), "context canceled") {
					return
				}
				errorCount.Add(1)
				t.Logf("Scheduler %d error: %v", schedulerID, err)
			},
			NextDelay:    10 * time.Millisecond,  // Fast polling
			LockDuration: 30 * time.Second,       // Reasonable lock time
		})
		if err != nil {
			t.Fatalf("Failed to create scheduler: %v", err)
		}

		schedulers = append(schedulers, sched)
	}

	// Start all schedulers simultaneously for maximum concurrency
	startMutex.Lock()
	for i, sched := range schedulers {
		wg.Add(1)
		go func(idx int, s *scheduler.Scheduler) {
			defer wg.Done()

			startMutex.Lock() // Wait for all goroutines to be ready
			startMutex.Unlock()

			if err := s.Start(ctx); err != nil {
				t.Errorf("Scheduler %d failed to start: %v", idx, err)
			}
		}(i, sched)
	}

	// Release all schedulers at once
	time.Sleep(100 * time.Millisecond) // Ensure all goroutines are waiting
	startMutex.Unlock()

	// Monitor progress
	ticker := time.NewTicker(5 * time.Second)
	defer ticker.Stop()

	done := make(chan struct{})
	go func() {
		wg.Wait()
		close(done)
	}()

	// Wait for all jobs to complete or timeout
	monitorCtx, monitorCancel := context.WithCancel(ctx)
	defer monitorCancel()

	go func() {
		for {
			select {
			case <-monitorCtx.Done():
				return
			case <-ticker.C:
				completed := executions.TotalExecutions()
				remaining, _ := countRemainingJobs(context.Background(), collection)
				t.Logf("Progress: %d/%d jobs executed, %d remaining, %d errors",
					completed, numJobs, remaining, errorCount.Load())
			}
		}
	}()

	// Wait for all jobs to be processed
	allProcessed := false
	checkInterval := 2 * time.Second
	maxIdleTime := 10 * time.Second
	lastProgress := executions.TotalExecutions()
	idleStart := time.Now()

	for !allProcessed {
		select {
		case <-ctx.Done():
			t.Fatal("Test timeout reached")
		case <-time.After(checkInterval):
			completed := executions.TotalExecutions()
			remaining, _ := countRemainingJobs(context.Background(), collection)

			// Check if we made progress
			if completed != lastProgress {
				lastProgress = completed
				idleStart = time.Now()
			}

			// If all jobs processed or been idle too long
			if remaining == 0 && completed >= numJobs {
				allProcessed = true
			} else if time.Since(idleStart) > maxIdleTime && remaining == 0 {
				// No progress and no jobs remaining
				allProcessed = true
			}
		}
	}

	monitorCancel()
	duration := time.Since(startTime)

	// Stop all schedulers
	t.Log("Stopping schedulers...")
	// Set stopping flag to ignore context.Canceled errors during shutdown.
	// These occur when a scheduler is mid-MongoDB call when its context gets canceled.
	// This is expected behavior and not an error - the schedulers stop cleanly.
	stopping.Store(true)
	stopCtx, stopCancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer stopCancel()

	for i, sched := range schedulers {
		if err := sched.Stop(stopCtx); err != nil {
			t.Logf("Warning: Scheduler %d stop error: %v", i, err)
		}
	}

	wg.Wait()

	// Analyze results
	separator := strings.Repeat("=", 80)
	t.Log("\n" + separator)
	t.Log("TEST RESULTS")
	t.Log(separator)

	stats := executions.GetStats()

	t.Logf("\nExecution Statistics:")
	t.Logf("  Total jobs queued:        %d", numJobs)
	t.Logf("  Total executions:         %d", stats.TotalExecutions)
	t.Logf("  Unique jobs executed:     %d", stats.UniqueJobs)
	t.Logf("  Jobs with duplicates:     %d", stats.DuplicateJobs)
	t.Logf("  Total duplicate runs:     %d", stats.TotalDuplicates)
	t.Logf("  Jobs not executed:        %d", stats.MissedJobs)
	t.Logf("  Errors encountered:       %d", errorCount.Load())

	t.Logf("\nPerformance Metrics:")
	t.Logf("  Total duration:           %v", duration)
	t.Logf("  Jobs per second:          %.2f", float64(stats.TotalExecutions)/duration.Seconds())
	t.Logf("  Avg time per job:         %v", duration/time.Duration(stats.TotalExecutions))

	// Show duplicate examples if any
	if len(stats.Duplicates) > 0 {
		t.Logf("\nDuplicate Executions Detected:")
		count := 0
		for jobID, execCount := range stats.Duplicates {
			if count < 10 { // Show first 10
				times := executions.GetTimestamps(jobID)
				t.Logf("  Job %s: executed %d times", jobID, execCount)
				for i, ts := range times {
					t.Logf("    Execution %d: %v", i+1, ts.Format(time.RFC3339Nano))
				}
			}
			count++
		}
		if count > 10 {
			t.Logf("  ... and %d more jobs with duplicates", count-10)
		}
	}

	// Show missed jobs if any
	if len(stats.MissedJobIDs) > 0 {
		t.Logf("\nMissed Jobs:")
		for i, jobID := range stats.MissedJobIDs {
			if i < 10 {
				t.Logf("  %s", jobID)
			}
		}
		if len(stats.MissedJobIDs) > 10 {
			t.Logf("  ... and %d more missed jobs", len(stats.MissedJobIDs)-10)
		}
	}

	t.Log(separator)

	// Validate results
	if stats.TotalDuplicates > 0 {
		t.Errorf("FAILED: Found %d duplicate executions across %d jobs",
			stats.TotalDuplicates, stats.DuplicateJobs)
	}

	if stats.MissedJobs > 0 {
		t.Errorf("FAILED: %d jobs were not executed", stats.MissedJobs)
	}

	if stats.UniqueJobs != numJobs {
		t.Errorf("FAILED: Expected %d unique jobs executed, got %d",
			numJobs, stats.UniqueJobs)
	}

	if stats.TotalDuplicates == 0 && stats.MissedJobs == 0 && stats.UniqueJobs == numJobs {
		t.Logf("\n✓ SUCCESS: All %d jobs executed exactly once with no duplicates!", numJobs)
		t.Logf("✓ Distributed locking mechanism is working correctly under high concurrency")
	}
}

// ExecutionTracker tracks job executions in a thread-safe manner
type ExecutionTracker struct {
	mu         sync.RWMutex
	counts     map[string]int
	timestamps map[string][]time.Time
}

func (et *ExecutionTracker) Record(jobID string) {
	et.mu.Lock()
	defer et.mu.Unlock()

	et.counts[jobID]++
	et.timestamps[jobID] = append(et.timestamps[jobID], time.Now())
}

func (et *ExecutionTracker) TotalExecutions() int {
	et.mu.RLock()
	defer et.mu.RUnlock()

	total := 0
	for _, count := range et.counts {
		total += count
	}
	return total
}

func (et *ExecutionTracker) GetTimestamps(jobID string) []time.Time {
	et.mu.RLock()
	defer et.mu.RUnlock()

	return et.timestamps[jobID]
}

type ExecutionStats struct {
	TotalExecutions int
	UniqueJobs      int
	DuplicateJobs   int
	TotalDuplicates int
	MissedJobs      int
	Duplicates      map[string]int
	MissedJobIDs    []string
}

func (et *ExecutionTracker) GetStats() ExecutionStats {
	et.mu.RLock()
	defer et.mu.RUnlock()

	stats := ExecutionStats{
		Duplicates:   make(map[string]int),
		MissedJobIDs: make([]string, 0),
	}

	stats.UniqueJobs = len(et.counts)

	for jobID, count := range et.counts {
		stats.TotalExecutions += count

		if count > 1 {
			stats.DuplicateJobs++
			stats.TotalDuplicates += (count - 1)
			stats.Duplicates[jobID] = count
		}
	}

	return stats
}

// Helper functions

func insertJobs(ctx context.Context, collection *mongo.Collection, count int) error {
	now := time.Now()

	// Insert in batches for better performance
	batchSize := 1000
	for i := 0; i < count; i += batchSize {
		batch := make([]interface{}, 0, batchSize)
		end := i + batchSize
		if end > count {
			end = count
		}

		for j := i; j < end; j++ {
			batch = append(batch, bson.M{
				"jobID":      fmt.Sprintf("job-%06d", j),
				"sleepUntil": now,
				"data":       fmt.Sprintf("test-job-%d", j),
			})
		}

		if _, err := collection.InsertMany(ctx, batch); err != nil {
			return fmt.Errorf("batch insert failed: %w", err)
		}
	}

	return nil
}

func countRemainingJobs(ctx context.Context, collection *mongo.Collection) (int64, error) {
	count, err := collection.CountDocuments(ctx, bson.M{
		"sleepUntil": bson.M{"$ne": nil},
	})
	return count, err
}

func getJobID(job *scheduler.Job) string {
	if jobID, ok := job.Data["jobID"].(string); ok {
		return jobID
	}
	return fmt.Sprintf("%v", job.ID)
}

File: mongodb-cron/src/scripts/concurrency-test.ts
import { MongoClient } from 'mongodb';
import { MongoCron } from '..';

const NUM_CRONS = 100;
const NUM_JOBS = 10000;

interface ExecutionTracker {
  counts: Map<string, number>;
  timestamps: Map<string, number[]>;
}

function createTracker(): ExecutionTracker {
  return {
    counts: new Map(),
    timestamps: new Map(),
  };
}

function recordExecution(tracker: ExecutionTracker, jobId: string) {
  const count = tracker.counts.get(jobId) || 0;
  tracker.counts.set(jobId, count + 1);

  const times = tracker.timestamps.get(jobId) || [];
  times.push(Date.now());
  tracker.timestamps.set(jobId, times);
}

function pad(num: number, size: number): string {
  return String(num).padStart(size, '0');
}

function analyzeResults(tracker: ExecutionTracker, numJobs: number) {
  const stats = {
    totalExecutions: 0,
    uniqueJobs: tracker.counts.size,
    duplicateJobs: 0,
    totalDuplicates: 0,
    missedJobs: 0,
    duplicates: new Map<string, number>(),
    missedJobIds: [] as string[],
  };

  for (const [jobId, count] of tracker.counts.entries()) {
    stats.totalExecutions += count;
    if (count > 1) {
      stats.duplicateJobs++;
      stats.totalDuplicates += (count - 1);
      stats.duplicates.set(jobId, count);
    }
  }

  const executedSet = new Set(tracker.counts.keys());
  for (let i = 0; i < numJobs; i++) {
    const jobId = 'job-' + pad(i, 6);
    if (!executedSet.has(jobId)) {
      stats.missedJobIds.push(jobId);
    }
  }
  stats.missedJobs = stats.missedJobIds.length;

  return stats;
}

async function runConcurrencyTest() {
  console.log('================================================================================');
  console.log('NODEJS MONGODB-CRON CONCURRENCY TEST');
  console.log('================================================================================');
  console.log('Configuration: ' + NUM_CRONS + ' crons, ' + NUM_JOBS + ' jobs');
  console.log('');

  const client = await MongoClient.connect('mongodb://localhost:27017');
  const dbName = 'scheduler_nodejs_concurrency_' + Date.now();
  const testDb = client.db(dbName);
  const collection = testDb.collection('jobs');

  try {
    console.log('Creating index...');
    await collection.createIndex({ sleepUntil: 1 }, { sparse: true });

    console.log('Inserting ' + NUM_JOBS + ' jobs...');
    const insertStart = Date.now();

    const jobs = [];
    for (let i = 0; i < NUM_JOBS; i++) {
      jobs.push({
        jobID: 'job-' + pad(i, 6),
        sleepUntil: new Date(),
        data: 'test-job-' + i,
      });
    }

    const batchSize = 1000;
    for (let i = 0; i < jobs.length; i += batchSize) {
      const batch = jobs.slice(i, i + batchSize);
      await collection.insertMany(batch);
    }

    const insertDuration = Date.now() - insertStart;
    console.log('Inserted ' + NUM_JOBS + ' jobs in ' + insertDuration + 'ms');
    console.log('');

    const tracker = createTracker();
    let errorCount = 0;
    let stopping = false;

    console.log('Starting ' + NUM_CRONS + ' concurrent MongoCron instances...');
    const startTime = Date.now();

    const crons: MongoCron[] = [];
    const startPromises: Promise<void>[] = [];

    for (let i = 0; i < NUM_CRONS; i++) {
      const cronId = i;

      const cron = new MongoCron({
        collection,
        onDocument: async (doc) => {
          const jobId = doc.jobID || ('unknown-' + doc._id);
          recordExecution(tracker, jobId);
        },
        onError: (err) => {
          if (stopping && err.message && err.message.includes('cursor killed')) {
            return;
          }
          errorCount++;
          if (errorCount <= 5) {
            console.log('Cron ' + cronId + ' error:', err.message);
          }
        },
        nextDelay: 10,
        reprocessDelay: 0,
        idleDelay: 0,
        lockDuration: 30000,
      });

      crons.push(cron);
    }

    for (const cron of crons) {
      startPromises.push(cron.start());
    }
    await Promise.all(startPromises);

    const progressInterval = setInterval(() => {
      const completed = tracker.counts.size;
      const executions = Array.from(tracker.counts.values()).reduce((sum, count) => sum + count, 0);
      console.log('Progress: ' + executions + ' executions, ' + completed + ' unique jobs');
    }, 5000);

    let allProcessed = false;
    let lastProgress = 0;
    let idleStartTime = Date.now();
    const maxIdleTime = 10000;

    while (!allProcessed) {
      await new Promise(resolve => setTimeout(resolve, 2000));

      const completed = tracker.counts.size;
      const remaining = await collection.countDocuments({ sleepUntil: { $ne: null } });

      if (completed !== lastProgress) {
        lastProgress = completed;
        idleStartTime = Date.now();
      }

      if (remaining === 0 && completed >= NUM_JOBS) {
        allProcessed = true;
      } else if (Date.now() - idleStartTime > maxIdleTime && remaining === 0) {
        allProcessed = true;
      }
    }

    clearInterval(progressInterval);
    const duration = Date.now() - startTime;

    console.log('Stopping all crons...');
    stopping = true;
    const stopPromises = crons.map(cron => cron.stop());
    await Promise.all(stopPromises);

    console.log('');
    console.log('================================================================================');
    console.log('TEST RESULTS');
    console.log('================================================================================');

    const stats = analyzeResults(tracker, NUM_JOBS);

    console.log('');
    console.log('Test Configuration:');
    console.log('  Concurrent crons:         ' + NUM_CRONS);
    console.log('  Total jobs queued:        ' + NUM_JOBS);

    console.log('');
    console.log('Execution Statistics:');
    console.log('  Total executions:         ' + stats.totalExecutions);
    console.log('  Unique jobs executed:     ' + stats.uniqueJobs);
    console.log('  Jobs with duplicates:     ' + stats.duplicateJobs);
    console.log('  Total duplicate runs:     ' + stats.totalDuplicates);
    console.log('  Jobs not executed:        ' + stats.missedJobs);
    console.log('  Errors encountered:       ' + errorCount);

    console.log('');
    console.log('Performance Metrics:');
    console.log('  Total duration:           ' + (duration / 1000).toFixed(3) + 's');
    console.log('  Jobs per second:          ' + (stats.totalExecutions / (duration / 1000)).toFixed(2));
    console.log('  Avg time per job:         ' + (duration * 1000 / stats.totalExecutions).toFixed(0) + 'us');
    console.log('  Throughput per cron:      ' + (stats.totalExecutions / (duration / 1000) / NUM_CRONS).toFixed(2) + ' jobs/sec');

    if (stats.duplicates.size > 0) {
      console.log('');
      console.log('WARNING: Duplicate Executions Detected!');
      console.log('Showing first 5 examples:');

      let count = 0;
      for (const [jobId, execCount] of stats.duplicates.entries()) {
        if (count < 5) {
          const times = tracker.timestamps.get(jobId) || [];
          console.log('');
          console.log('  Job ' + jobId + ': executed ' + execCount + ' times');
          times.forEach((time, i) => {
            console.log('    Execution ' + (i + 1) + ': ' + new Date(time).toISOString());
            if (i > 0) {
              const gap = time - times[i - 1];
              console.log('      Time gap: ' + gap + 'ms');
            }
          });
        }
        count++;
      }

      if (count > 5) {
        console.log('  ... and ' + (count - 5) + ' more jobs with duplicates');
      }
    }

    if (stats.missedJobIds.length > 0) {
      console.log('');
      console.log('Missed Jobs:');
      stats.missedJobIds.slice(0, 10).forEach(jobId => {
        console.log('  ' + jobId);
      });
      if (stats.missedJobIds.length > 10) {
        console.log('  ... and ' + (stats.missedJobIds.length - 10) + ' more missed jobs');
      }
    }

    console.log('');
    console.log('================================================================================');

    if (stats.totalDuplicates > 0) {
      console.log('FAILED: Found ' + stats.totalDuplicates + ' duplicate executions across ' + stats.duplicateJobs + ' jobs');
      process.exit(1);
    }

    if (stats.missedJobs > 0) {
      console.log('FAILED: ' + stats.missedJobs + ' jobs were not executed');
      process.exit(1);
    }

    if (stats.uniqueJobs !== NUM_JOBS) {
      console.log('FAILED: Expected ' + NUM_JOBS + ' unique jobs executed, got ' + stats.uniqueJobs);
      process.exit(1);
    }

    console.log('SUCCESS: All ' + NUM_JOBS + ' jobs executed exactly once!');
    console.log('NO DUPLICATES found with ' + NUM_CRONS + ' concurrent crons!');
    console.log('Distributed locking mechanism PASSED stress test!');
    console.log('Achieved ' + (stats.totalExecutions / (duration / 1000)).toFixed(0) + ' jobs/second throughput');
    console.log('');

  } finally {
    await testDb.dropDatabase();
    await client.close();
  }
}

runConcurrencyTest().catch((error) => {
  console.error('Test failed with error:', error);
  process.exit(1);
});
